src/front/glsl/builtins.rs

```diff
@@ -165,184 +237,79 @@ pub fn inject_builtin(declaration: &mut FunctionDeclaration, module: &mut Module
                     class,
                 };
 
-                let vector = match (dim, shadow) {
-                    (0b00, false) => TypeInner::Scalar {
-                        kind: Sk::Float,
-                        width,
-                    },
-                    (0b00, true) | (0b11, _) => TypeInner::Vector {
-                        size: VectorSize::Tri,
-                        kind: Sk::Float,
-                        width,
-                    },
-                    (_, _) => {
-                        let size = match dim + arrayed as u32 + shadow as u32 {
-                            1 => VectorSize::Bi,
-                            2 => VectorSize::Tri,
-                            _ => VectorSize::Quad,
-                        };
-
-                        TypeInner::Vector {
-                            size,
-                            kind: Sk::Float,
-                            width,
-                        }
+                let num_coords_from_dim = (dim + 1).min(3);
+                let mut num_coords = num_coords_from_dim;
+
+                if shadow && proj {
+                    num_coords = 4;
+                } else if shadow {
+                    num_coords += 1;
+                } else if proj {
+                    if variant && num_coords == 4 {
+                        // Normal form already has 4 components, no need to have a variant form.
+                        continue;
+                    } else if variant {
+                        num_coords = 4;
+                    } else {
+                        num_coords += 1;
```

Comment:
nit: When dealing with this kind of `else if` with the same variables I like to use a `match` because it looks like a truth table, something like:

```rust
match (shadow, proj) {
  (true, true) => num_coords = 4,
  (true, false) => num_coords += 1,
  (false, true) => ...,
  (false, false) => {}
}
```

Reply:
I find this much more difficult to read.

Reply:
understandable, ignore this then